x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/IndicesPermission.java

```diff
@@ -656,6 +672,58 @@ private static boolean containsPrivilegeThatGrantsMappingUpdatesForBwc(Group gro
         return group.privilege().name().stream().anyMatch(PRIVILEGE_NAME_SET_BWC_ALLOW_MAPPING_UPDATE::contains);
     }
 
+    /**
+     * Get all automatons for the index groups in this permission and optionally combine the index groups to enable checking if a set of
+     * index patterns specified using a regular expression grants a set of index privileges.
+     *
+     * <p>An index group is defined as a set of index patterns and a set of privileges (excluding field permissions and DLS queries).
+     * {@link IndicesPermission} consist of a set of index groups. For non-regular expression privilege checks, an index pattern is checked
+     * against each index group, to see if it's a sub-pattern of the index pattern for the group and then if that group grants some or all
+     * of the privileges requested. For regular expressions it's not sufficient to check per group since the index patterns covered by a
+     * group can be distinct sets and a regular expression can cover several distinct sets.
+     *
+     * <p>For example the two index groups: {"names": ["a"], "privileges": ["read", "create"]} and {"names": ["b"],
+     * "privileges": ["read","delete"]} will not match on ["\[ab]\"], while a single index group:
+     * {"names": ["a", "b"], "privileges": ["read"]} will. This happens because the index groups are evaluated against a request index
+     * pattern without first being combined. In the example above, the two index patterns should be combined to:
+     * {"names": ["a", "b"], "privileges": ["read"]} before being checked.
+     *
+     *
+     * @param combine combine index groups to allow for checking against regular expressions
+     *
+     * @return a list of tuples of all index and privilege pattern automaton
+     */
```

Comment:
Did some more testing on this, since the number of automatons grow exponentially we can't support more than ~10 index groups. I'll work on optimizing this.

Reply:
Did some optimizations: 
- Only combine when we need to match regexp 
- Use a map of privilege automaton -> index pattern automaton. Automatons do not have a hashcode or equals because that would be inefficient. The map here works a a cache that relies on automaton object references being equal. This works because they're cached in `IndexPrivilege::CACHE`. 
- The current implementation works well as long as you don't have a lot of index groups that contain a distinct list of privileges since the cache can't be used then. 

Since the performance concerns are now limited to only when there is regexp matching combined with a lot of index groups where each group specified more than one privilege, I think this is probably acceptable.